dxp/sennajs_integration.md

```diff
@@ -0,0 +1,117 @@
+
+The `frontend-js-spa-web` modules also contains templates to [configure](https://github.com/liferay/liferay-portal/blob/2d4120b7fdb8cf03e7505ebdaacee3bdceec3391/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/config.tmpl) and [initialize](https://github.com/liferay/liferay-portal/blob/2d4120b7fdb8cf03e7505ebdaacee3bdceec3391/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/init.tmpl) the senna.js `App`.
+
+Everything is included with the [SPATopHeadJSPDynamicInclude](https://github.com/liferay/liferay-portal/blob/2d4120b7fdb8cf03e7505ebdaacee3bdceec3391/modules/apps/frontend-js/frontend-js-spa-web/src/main/java/com/liferay/frontend/js/spa/web/internal/servlet/taglib/SPATopHeadJSPDynamicInclude.java#L134-L143) if DXP is configured to use SPA, which it is by default (except in [Internet Explorer](https://github.com/liferay/liferay-portal/blob/454d7f6a439dc668a0ce0529cff2cae87ee9a0d5/modules/apps/frontend-js/frontend-js-spa-web/src/main/java/com/liferay/frontend/js/spa/web/internal/configuration/SPAConfiguration.java#L38-L50)).
+
+Disabling SPA in DXP can be done by setting adding the `javascript.single.page.application.enabled=false` key/value pair to your `portal-ext.properties` file.
+Additionally, if you want to enable Single Application Mode for Internet Explorer, you can do so by navigating to: **Control Panel** > **System Settings** > **Infrastructure** > **Frontend SPA Infrastructure** and unchecking the checkboxes for Internet Explorer, but keep in mind that this might change in the future and that it is not recommended.
+
+## Link selectors
+
+By default, the Senna.js `App` instance will handle all click events on links (`<a>` tags), that match the CSS selectors specified in the `Liferay.SPA.navigationExceptionSelectors` global variable.
+
+These CSS selectors can also be configured by navigating to **Control Panel** > **System Settings** > **Infrastructure** > **Frontend SPA Infrastructure** and changing the values specificed in the **Navigation Exception Selectors** fields.
+
+## Form submission
+
+The form submission will also be handled by the Senna.js `App` instance, unless the form's `action` attribute is set to `get`
+
+## Block list
+
+It's also possible to tell the Senna.js `App` instance which portlets it needs to ignore, in which case it will not be used for navigation or form submission.
+
+Any portlet that is set to not use "Single Application Mode", or any portlet that is not deployed (and not ready) is automatically added to this list.
+
+## Cache
+
+The Senna.js `App` will use caching, if the `Liferay.SPA.cacheExpirationTime` variable's value is greater than `-1`.
```

Comment:
Really? So if I set it to `-0.9` it will still cache?

Reply:
It [looks](https://github.com/liferay/liferay-portal/blob/021b09a2df5e71140dd6061497000db108f09048/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/liferay/app/App.es.js#L136-L138) like it will

Reply:
Although [here](https://github.com/liferay/liferay-portal/blob/021b09a2df5e71140dd6061497000db108f09048/modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/liferay/app/App.es.js#L171-L177), it looks like if you set a negative time `screenCachedExpired` will always be `true` (because `lastModifiedInterval` should always be non-negative), which is the same as having no cache.

So technically, unless I misunderstand how this works, the actual truth is that Senna will cache if the value is greater than `0`, not `-1`.

🤷

Reply:
Oh ... right.

Reply:
I'll fix that